File: src/preread/response_parser.py
"""
src/preread/response_parser.py
------------------------------
Responsible for one thing: parsing the structured text response from the
preread API call into a dict of section name → content string.

This module has no knowledge of the API, file I/O, or prompt construction.
It receives a raw string and returns a plain dict. Nothing more.

To change the expected output format (section headers), edit only this file.
"""

import re

# The five section headers the model is instructed to produce, in order.
# The key is the canonical section name used as a dict key downstream.
# The value is the exact header string the model produces.
SECTION_HEADERS = {
    "characters":       "=== CHARACTERS ===",
    "locations":        "=== LOCATIONS ===",
    "terminology":      "=== TERMINOLOGY ===",
    "cultural_phrases": "=== CULTURAL PHRASES ===",
    "story":            "=== STORY ===",
}


def parse_response(raw: str) -> dict[str, str]:
    """
    Parse a structured preread API response into a dict of section contents.

    Parameters
    ----------
    raw : str
        The full text response from the API.

    Returns
    -------
    dict[str, str]
        Keys: canonical section names (see SECTION_HEADERS).
        Values: content of that section (stripped), or empty string if the
                section was absent or contained only "NOTHING TO ADD".

    Notes
    -----
    - Sections not found in the response are returned as empty strings,
      with a warning printed to stdout.
    - "NOTHING TO ADD" content is normalized to an empty string.
    """
    results: dict[str, str] = {key: "" for key in SECTION_HEADERS}

    # Build a pattern that splits on any of our known headers.
    header_pattern = "|".join(re.escape(h) for h in SECTION_HEADERS.values())
    # We split on the headers themselves, keeping them as delimiters.
    parts = re.split(f"({header_pattern})", raw)

    # parts alternates: [pre-text, header, content, header, content, ...]
    # Walk through pairs of (header, content).
    i = 1
    while i < len(parts) - 1:
        header_text = parts[i].strip()
        content_text = parts[i + 1].strip() if i + 1 < len(parts) else ""
        i += 2

        # Find which canonical key this header maps to.
        matched_key = next(
            (key for key, h in SECTION_HEADERS.items() if h == header_text),
            None
        )
        if matched_key is None:
            continue

        # Normalize "nothing to add" to empty string.
        if content_text.upper().strip() == "NOTHING TO ADD":
            content_text = ""

        results[matched_key] = content_text

    # Warn about any sections the model failed to produce.
    for key, header in SECTION_HEADERS.items():
        if key not in results or results[key] == "":
            pass  # Silence is fine — empty means nothing to add.

    missing = [h for key, h in SECTION_HEADERS.items()
               if key not in results]
    if missing:
        print(f"  [warning] Response missing expected sections: {missing}")

    return results

File: app/services/pipeline/ruby/preread_runner/response_parser.rb
# ---------------------------------------------------------------------------
# Pipeline::Ruby::PrereadRunner::ResponseParser
#
# Ruby port of src/preread/response_parser.py — splits the model's structured
# preread response into the five canonical sections. No knowledge of the
# API, file I/O, or prompt construction.
#
# One deliberate hardening beyond the Python original (per
# docs/RAILS_REFACTOR_PLAN.md's R6 acceptance criteria, same fail-closed
# discipline R5 established): Python silently treats a response with none of
# the five headers the same as "every section legitimately empty" — its own
# "missing sections" warning is dead code, since results is pre-populated
# with every key before the check runs. This port distinguishes the two: a
# response with zero section markers at all is a hard parse failure the
# caller must check for (result.success?), never silently equivalent to a
# well-formed "nothing to add" response.
# ---------------------------------------------------------------------------
module Pipeline
  module Ruby
    class PrereadRunner
      module ResponseParser
        SECTION_HEADERS = {
          characters:        "=== CHARACTERS ===",
          locations:         "=== LOCATIONS ===",
          terminology:       "=== TERMINOLOGY ===",
          cultural_phrases:  "=== CULTURAL PHRASES ===",
          story:             "=== STORY ==="
        }.freeze

        HEADER_PATTERN = /(#{SECTION_HEADERS.values.map { |h| Regexp.escape(h) }.join('|')})/

        Result = Struct.new(:sections, :missing_markers, keyword_init: true) do
          def success?
            !missing_markers
          end
        end

        def self.parse(raw)
          return Result.new(sections: {}, missing_markers: true) if SECTION_HEADERS.values.none? { |h| raw.include?(h) }

          sections = SECTION_HEADERS.keys.index_with { "" }
          parts = raw.split(HEADER_PATTERN)

          i = 1
          while i < parts.length - 1
            header_text  = parts[i].strip
            content_text = parts[i + 1] ? parts[i + 1].strip : ""
            i += 2

            key = SECTION_HEADERS.key(header_text)
            next unless key

            content_text = "" if content_text.strip.upcase == "NOTHING TO ADD"
            sections[key] = content_text
          end

          Result.new(sections: sections, missing_markers: false)
        end
      end
    end
  end
end
